Adding augmentation utilities for AudioSignal objects

abugler opened the pull request:
This pull request adds `augmentation.py`, which includes augmentation utilities for AudioSignal objects, using ffmpeg. This includes:
 + Tremolo
 + Vibrato
 + Low/High pass
 + Chorus
 + Flanger
 + Phaser
 + Emphasis
 + Compression
 + Equalization
 + Pitch shifting
 + Time Stretching

pseeth commented on nussl/core/augmentation.py:
```diff
@@ -0,0 +1,377 @@
+# that use the level_in arugment, however, not all filters that
+# use this state these bounds in documentationwith pytest.raises(ValueError):
+LEVEL_MIN = .015625
+LEVEL_MAX = 64
```
These should be in `constants.py`.

pseeth commented on nussl/core/augmentation.py:
```diff
@@ -0,0 +1,377 @@
+    # stretched_signal = audio_signal.make_copy_with_audio_data(stretched_audio_data)
+
+    # This one does not
+    stretched_signal = AudioSignal(audio_data_array=stretched_audio_data)
```
I think we should use the one that throws a warning. It's more clear. We can decide whether to filter the warning later. But right now, the STFT parameters will not be shared if you do it this way.

pseeth commented on nussl/core/augmentation.py:
```diff
@@ -0,0 +1,377 @@
+
+    # This one does not
+    stretched_signal = AudioSignal(audio_data_array=stretched_audio_data)
+    stretched_signal.stft()
```
Why do we need to take the STFT here?

abugler replied:
If you make a copy, do you not have to take the STFT? I thought that did not automatically populate?

pseeth replied:
The user should take the STFT instead, otherwise we are doing some side computation that the user didn't request and may not need. Function should only do what is specified.

pseeth commented on nussl/core/augmentation.py:
```diff
@@ -0,0 +1,377 @@
+        shifted_audio_data.append(librosa.effects.pitch_shift(audio_row, sample_rate, shift))
+    shifted_audio_data = np.array(shifted_audio_data)
+    shifted_signal = audio_signal.make_copy_with_audio_data(shifted_audio_data)
+    shifted_signal.stft()
```
Like before, why take the STFT?

pseeth commented on nussl/core/augmentation.py:
```diff
@@ -0,0 +1,377 @@
+        augmented_signal: A copy of the original audio signal, with augmentations applied 
+    """
+    if not np.issubdtype(type(lowest_freq), np.number) or lowest_freq <= 0:
+        raise ValueError("lowest_freq should be positve scalar")
```
positve -> positive

pseeth commented on nussl/core/augmentation.py:
```diff
@@ -0,0 +1,377 @@
+    if not np.issubdtype(type(mod_depth), np.number) or mod_depth < 0 or mod_depth > 1:
+        raise ValueError("mod_depth should be positve scalar between 0 and 1.")
+
+    filter_kwargs = {
```
Why not do this directly? Why create a kwargs dict first then use it?

abugler replied:
So for every ffmpeg filter, I found myself writing the code found in lines 52-71 in `augmentation_utils.py` in every single function. The only difference in each filter is the kwargs for `ffmpeg.filter()`. So this became the primary input to `apply_ffmpeg_filter`. If I was not to use kwargs here, I would have to declare every single possible filter argument in `apply_ffmpeg_filter`.

pseeth replied:
Instead of making a dictionary, you can pass them as keyword arguments to `apply_ffmpeg_filter`, instead of making a dictionary and then passing the dictionary.

abugler replied:
Just made that change in my code. I didn't actually know you could do that.

pseeth commented on nussl/core/augmentation.py:
```diff
@@ -0,0 +1,377 @@
+    if not np.issubdtype(type(mod_depth), np.number) or mod_depth < 0 or mod_depth > 1:
+        raise ValueError("mod_depth should be positve scalar between 0 and 1.")
+
+    filter_kwargs = {
```
kwargs dict?

pseeth commented on nussl/core/augmentation.py:
```diff
@@ -0,0 +1,377 @@
+    """
+
+    # TODO: Arg checks
+    filter_kwargs = {
```
kwargs dict?

pseeth commented on nussl/core/augmentation.py:
```diff
@@ -0,0 +1,377 @@
+    filter_kwargs = {
+        "in_gain": in_gain,
+        "out_gain": out_gain,
+        "delays": make_arglist_ffmpeg(delays),
```
I think these utilities should go into this folder, instead of into `augmentation_utils.py`.

pseeth commented on nussl/core/augmentation.py:
```diff
@@ -0,0 +1,377 @@
+        augmented_signal: A copy of the original audio signal, with augmentations applied
+    """
+    # TODO: Arg checks
+    filter_kwargs = {
```
kwargs dict

pseeth commented on nussl/core/augmentation.py:
```diff
@@ -0,0 +1,377 @@
+        augmented_signal: A copy of the original audio signal, with augmentations applied
+    """
+    # TODO: Arg checks
+    filter_kwargs = {
```
kwargs dict

pseeth commented on nussl/core/augmentation.py:
```diff
@@ -0,0 +1,377 @@
+    Returns:
+        augmented_signal: A copy of the original audio signal, with augmentations applied 
+    """
+    if level_in < LEVEL_MIN or level_in > LEVEL_MAX \
```
`constants.LEVEL_MIN/LEVEL_MAX` once you make the changes.

ethman commented on nussl/core/augmentation.py:
```diff
@@ -0,0 +1,377 @@
+
+    return apply_ffmpeg_filter(audio_signal, "vibrato", **filter_kwargs)
+
+def chorus(audio_signal, in_gain=.4, out_gain=.4, delays=[50], 
```
```suggestion
def chorus(audio_signal, in_gain=.4, out_gain=.4, delays=[50], 
decays=[.8], speeds=[.95], depths=[.7]):
```
Don't make default arguments mutable: https://docs.python-guide.org/writing/gotchas/#mutable-default-arguments

pseeth replied:
Simple fix here is to do `delays=None`, then take care of it in the body of the function `delays = delays if delays is not None else [50]`.

pseeth commented on nussl/core/augmentation.py:
```diff
@@ -0,0 +1,377 @@
+    detection, mix, threshold):
+
+    # The following values are taken from ffmpeg documentation
+    if level_in < LEVEL_MIN or level_in > LEVEL_MAX or \
```
Use something more like

```
if (
   level_in < LEVEL_MIN or level_in > LEVEL_MAX or
   mode not in {"upward", "downward"} or 
)
```

Parens instead of \ for line breaks.

pseeth commented on nussl/core/augmentation.py:
```diff
@@ -0,0 +1,377 @@
+    """
+    # TODO: Argcheck
+
+    filter_kwargs = {
```
I can see why you would use kwargs here.

pseeth commented on nussl/core/augmentation_utils.py:
```diff
@@ -0,0 +1,72 @@
+from .audio_signal import AudioSignal
+import ffmpeg
+
+def save_audio_signal_to_tempfile(audio_signal, extension=".wav"):
```
We should use a different style for temporary files. See here for an example:

https://github.com/nussl/nussl/blob/master/nussl/core/play_utils.py#L147

Essentially, use a context manager to make sure everything is safe. This solution is also portable with Windows.

pseeth commented on nussl/core/augmentation_utils.py:
```diff
@@ -0,0 +1,72 @@
+silent_kwargs = {'loglevel': 'quiet'}
+
+def apply_ffmpeg_filter(audio_signal, _filter, silent=True, **kwargs):
+    audio_tempfile, audio_tempfile_name = \
```
Use the tempfiles solution from before. All of this can be replaced with a good context manager.

pseeth commented on tests/core/test_augmentation.py:
```diff
@@ -0,0 +1,231 @@
+import numpy as np
+import numpy.random as random
+from nussl.core.augmentation import *
```
Try not to import like this. Import explicitly, not implicitly.

pseeth commented on nussl/core/augmentation.py:
```diff
@@ -0,0 +1,377 @@
+import os
+from .audio_signal import AudioSignal
+import tempfile
+from .augmentation_utils import *
```
Import explicitly the functions you need. I think these utils should be moved into `augment.py` directly, actually.

pseeth commented on nussl/core/augmentation.py:
```diff
@@ -4,7 +4,7 @@
 from .audio_signal import AudioSignal
 import tempfile
-from .augmentation_utils import *
+from .augmentation_utils import apply_ffmpeg_filter
```
Let's move this into augment.py instead of keeping it separate.

abugler commented:
Changed `effects.py` such that all SoX effects use PySoX 1.4.0

ethman commented:
One last comment about the docs: as the API is getting a little more stable, think about adding some example code for each effect. The easier ones can be made in-line (see below), but the more involved ones we should link to an examples page in the docs.

Eg time stretch is easy:

    >>> audio_signal.signal_duration
    3.0
    >>> audio_signal.time_stretch(0.5)
    >>> audio_signal.signal_duration
    1.5

abugler commented:
In case it wasn't obvious by the last pull request, things I've changed:
 - I put some example code in `AudioSignal.build_effects`. There is also example code and usage notes at the stop of `effects.py`, since I didn't really see it to be necessary to put nearly the same  example code on every effect. 
 - Added argument types to many of the functions. 
 - Improved some Error handling

abugler commented:
Between commits 86779b3 and d8417a4, 4 more errors occur in unrelated files. These being `tests/ml/train_trainer.py`, `tests/ml/test_overfit.py`, and `tests/ml/tests_closures.py`. Unsure why this is happening, and will be investigating.

pseeth commented:
lgtm. @ethman?

ethman commented on nussl/core/audio_signal.py:
```diff
@@ -1644,6 +1649,311 @@ def embed_audio(self, ext='.mp3', display=True):
+    #          Effect Hooks          #
+    ##################################
+
+    def build_effects(self, reset=True, overwrite=False):
```
Does this apply effects to my audio or is this a preprocessing step before I can apply effects? If the former, this method should be renamed to `apply_effects`, if the latter then the documentation needs to be clearer.

abugler replied:
This applies effects. Changed function name

ethman commented on nussl/core/audio_signal.py:
```diff
@@ -1644,6 +1649,311 @@ def embed_audio(self, ext='.mp3', display=True):
+    def build_effects(self, reset=True, overwrite=False):
+        """
+        Builds all effects in the AudioSignal's effects chains using the audio data of the calling 
+        AudioSignal. If reset=True, the current effects chains will be deleted. If overwrite=True, 
```
What exactly does `reset=True` do? Does it clear the effects chain before the effects are applied if there are any left over from last time?

abugler replied:
Removes effects in effects chains after applying effects. Made this more clear in docs

ethman commented on nussl/core/audio_signal.py:
```diff
@@ -1644,6 +1649,311 @@ def embed_audio(self, ext='.mp3', display=True):
+            with edited audio_data. Otherwise, returns a new AudioSignal new_signal.
+        """
+        if not self._ffmpeg_effects_chain and not self._sox_effects_chain:
+            raise RuntimeError("No effect hooks have been called on this AudioSignal")
```
I don't think this should raise an error. If there are no effects, it should just `return self` without making a fuss.

abugler replied:
Removed Error

ethman commented on nussl/core/audio_signal.py:
```diff
@@ -1644,6 +1649,311 @@ def embed_audio(self, ext='.mp3', display=True):
+        If no effects have been added to the AudioSignal's effects chains, then build_effects 
+        will throw a RuntimeError. 
+
+        This will add a tremolo effect followed by a high pass effect to the AudioSignal's effects chains:
```
I think you need a few more examples here. For instance, if you start a very easy one like  `time_stretch`, you can show the length of the signal to demonstrate how it should work. E.g.,

`>>> my_signal.duration`
`10.0`
`>>> my_signal.time_stretch(0.5)`
`>>> my_signal.build_effects(overwrite=True)`
`>>> my_signal.duration`
`5.0`

Then you can demonstrate daisy chaining and other advanced things.

abugler replied:
Added Example in docs

ethman commented on nussl/core/audio_signal.py:
```diff
@@ -1644,6 +1649,311 @@ def embed_audio(self, ext='.mp3', display=True):
+
+        return new_signal
+        
+    def make_effect(self, effect, **params):
```
What does this function do?

abugler replied:
This function allows for the user to call a certain effect by passing the name of the effect as a string. I found this function very convenient when trying to make OnTheFly datasets.

abugler replied:
changed docs to make it more clear

ethman commented on nussl/core/audio_signal.py:
```diff
@@ -1644,6 +1649,311 @@ def embed_audio(self, ext='.mp3', display=True):
+        Calls another effect hook.
+        Args:
+            effect (str): Function name of desired effect hook of the AudioSignal
+            **params: Additional parameters for given effect. 
```
Change `**params` to `**kwargs`.

abugler replied:
Renamed variable

ethman commented on nussl/core/constants.py:
```diff
@@ -95,3 +95,10 @@
+# that use the level_in arugment, however, not all filters that
+# use this state these bounds in documentationwith pytest.raises(ValueError):
+LEVEL_MIN = .015625
+LEVEL_MAX = 64
```
PEP8: needs a newline at end of the file

abugler replied:
Fixed

ethman commented on nussl/core/effects.py:
```diff
@@ -0,0 +1,584 @@
+         .run()
+         )
+
+        augmented_signal = AudioSignal(path_to_input_file=out_tempfile.name)
```
This won't keep metadata from the input `audio_signal`. I'm pretty sure the only metadata that you'll want to keep is `audio_signal.label`. This will also set `augmented_signal`'s path to the tempfile name, instead of the path of the input audio signal. I'm torn about whether this is okay or not. What are your thoughts @pseeth ?

abugler replied:
I'll copy over in the `audio_signal.label` to the new signal. I think it would make sense if the `augmented_signal`'s path is just set to `None`.

ethman replied:
You should also copy over effects metadata (https://github.com/nussl/nussl/pull/162#commitcomment-39648466)

ethman commented on nussl/core/effects.py:
```diff
@@ -0,0 +1,584 @@
+    allowed_interp_types = {"linear", "quadratic"}
+    if interp not in allowed_interp_types:
+        raise ValueError(f"interp must be one of the following:\n{allowed_interp_types}")
+    if (delay < 0 or delay > 30
```
Two things here:
1) I think you might be able to just do `if not (0 < delay < 30):`
2) I think it might be worth splitting each of these in its own if statement and throwing a separate exception for each. Yes it's ugly, but this whole section is already pretty hairy, so we might as well make things easier for the user by telling them exactly what's wrong.

abugler replied:
Changed to print only values that are out of bounds

ethman commented on nussl/core/effects.py:
```diff
@@ -0,0 +1,584 @@
+                         attack, release, makeup, knee, link,
+                         detection, mix, threshold):
+    # The following values are taken from ffmpeg documentation
+    if (level_in < LEVEL_MIN or level_in > LEVEL_MAX or
```
Same comment as the flanger argcheck function

abugler replied:
Fixed

abugler commented:
Demo Notebook :) https://nbviewer.jupyter.org/github/abugler/nussl_effects_notebook/blob/master/effects_notebook.ipynb

ethman commented on nussl/core/audio_signal.py:
```diff
@@ -201,6 +201,7 @@ def __init__(self, path_to_input_file=None, audio_data_array=None, stft=None, la
         # Effects
         self._ffmpeg_effects_chain = []
         self._sox_effects_chain = []
+        self.effects_applied = []
```
Awesome thanks! This needs to documented now in the Attributes section (here: https://github.com/nussl/nussl/blob/ce3f8792dcfe10c28c2ba2b5a8c393580137adec/nussl/core/audio_signal.py#L159)

ethman replied:
Also add a reference to the explaining docs in `apply_effects()` ("for more info see...")

ethman commented on nussl/core/audio_signal.py:
```diff
@@ -1653,73 +1654,99 @@ def embed_audio(self, ext='.mp3', display=True):
-        Builds all effects in the AudioSignal's effects chains using the audio data of the calling 
-        AudioSignal. If reset=True, the current effects chains will be deleted. If overwrite=True, 
-        the calling AudioSignal's audio_data will be overwritten and the stft_data will be deleted. 
+        Applies all effects in the AudioSignal's effects chains using the audio data of the calling 
```
I think this needs a little more fleshing out here. There are a few concepts that are not explicitly stated that need to be. Here is the place for us to spell out how the whole effects process works, conceptually. Questions that need to be addressed here: What's an "effects chain"?  How do I put effects in the chain? How are the effects stored/how can I access it to see what's there? Does the order of elements in the chain matter? What steps do I need to take before they're applied? Say it with language first, then talk about the details (what `reset`/`overwrite` do, which are FFmpeg vs SoX, etc), then provide a few bits of example code.

Also, you'll need to add one line of documentation to every effect function saying something to the _effect_ ;-) of "this won't get applied until you call `apply_effects()` and the order matters. for more details see link:`apply_effects()`". It's fine to copy and paste the same text to every effect function both here and in effects.py

ethman commented on nussl/core/audio_signal.py:
```diff
@@ -1653,73 +1654,99 @@ def embed_audio(self, ext='.mp3', display=True):
 
-        >>> audio_signal.build_effects(overwrite=True)
+        All effects applied will stored in `self.effects_applied`, a list of FilterFunctions
+        in the order that they were applied.
```
I would also show the 1-liner version of this as icing on the cake: `new_signal = audio_signal.time_stretch(0.5).tremolo(5, .6).high_pass(12000).apply_effects()`

ethman commented on nussl/core/audio_signal.py:
```diff
@@ -1653,73 +1654,99 @@ def embed_audio(self, ext='.mp3', display=True):
     def reset_effects_chain(self):
         """
-        Clears effects chain of AudioSignal
+        Clears effects chains of AudioSignal
```
Add a line like "for more information on applying effects in AudioSignal see link:`apply_effects()`

ethman commented:
Awesome work! I added a few more comments. I think the final things for me are all docs related, as noted inline above, but I'll add a checklist here:

- [ ] Do a thorough documentation conceptually explaining the whole process in `apply_effects()`
- [ ] Document `self.effects_applied`
- [ ] Add your notebook to the tutorials in the docs
- [ ] Make sure links work, both to other parts of the docs and to external pages (ffmpeg docs links)

See here for instructions on updating docs: https://nussl.github.io/docs/contributing.html#documentation

I talked to @pseeth  and we're on the same page about all of this.

pseeth commented:
Awesome work. @ethman, this looks good to me to merge? After merge, I'll update PyPI and update the docs page.

pseeth commented:
Just fixing up some docs building stuff. The chorus effect sounds a bit weird still on my end. If you can fix that, we're good to go, as long as tests pass. I also added the ability to pass in arbitrary arguments to both `time_stretch` and `pitch_shift`.

ethman commented:
I was looking over these changes again and I was wondering why SoX effects have to get applied before ffmpeg effects. Is there a major reason why this is? Is it _that_ much slower to apply the effects one-by-one (perhaps there's a way to batch while preserving order, e.g. if there are 6 ffmpegs in a row then a SoX)? I think, above all, we should honor the user's ordering.

ethman commented:
In `AudioSignal.applied_effects` can we also store the effect kwargs? Shouldn't be too hard, just overload a `FilterFunction.__str__()`. You'll have to migrate `**filter_kwargs` to that class and store them in `self`. Seems like there might be some other shared functionality you can push to that superclass as well.

abugler commented:
> I was looking over these changes again and I was wondering why SoX effects have to get applied before ffmpeg effects. Is there a major reason why this is? Is it _that_ much slower to apply the effects one-by-one (perhaps there's a way to batch while preserving order, e.g. if there are 6 ffmpegs in a row then a SoX)? Do you have benchmarks for "SoX-then-ffmpeg" vs. "user order"?
> 
> I think, above all, we should honor the user's ordering. Is there an engineering reason that would make this difficult?

This is something that Prem and I decided on quite awhile ago. The issue with FFmpeg effects, is that the audio data must be stored in a tempfile before processing. Therefore, it would much more efficient to store the signal in a tempfile, then apply several ffmpeg effects, then read the tempfile containing the signal. 

If all SoX effects came after FFmpeg effects, (or vice versa) it wouldn't impact the speed too much. However, if the user wanted to do effects in the order FFmpeg-SoX-FFmpeg, it would present more of an issue, as it would require two tempfile reads. 

I don't think it would be too much of an issue to implement user ordering, but speed is something that we were trying to prioritize at the time. @pseeth what do you think about this?

ethman commented:
> > I was looking over these changes again and I was wondering why SoX effects have to get applied before ffmpeg effects. Is there a major reason why this is? Is it _that_ much slower to apply the effects one-by-one (perhaps there's a way to batch while preserving order, e.g. if there are 6 ffmpegs in a row then a SoX)? Do you have benchmarks for "SoX-then-ffmpeg" vs. "user order"?
> > I think, above all, we should honor the user's ordering. Is there an engineering reason that would make this difficult?
> 
> This is something that Prem and I decided on quite awhile ago. The issue with FFmpeg effects, is that the audio data must be stored in a tempfile before processing. Therefore, it would much more efficient to store the signal in a tempfile, then apply several ffmpeg effects, then read the tempfile containing the signal.
> 
> If all SoX effects came after FFmpeg effects, (or vice versa) it wouldn't impact the speed too much. However, if the user wanted to do effects in the order FFmpeg-SoX-FFmpeg, it would present more of an issue, as it would require two tempfile reads.
> 
> I don't think it would be too much of an issue to implement user ordering, but speed is something that we were trying to prioritize at the time. @pseeth what do you think about this?

Do you have any side-by-side speed tests? I think that would answer whether or not we should do it.

abugler commented:
@ethman Just ran some speed tests. 
![image](https://user-images.githubusercontent.com/28627542/87695619-5034b300-c755-11ea-96cd-9e1a03f4d129.png)

abugler commented:
With 1000 samples
![image](https://user-images.githubusercontent.com/28627542/87712202-f42a5880-c76d-11ea-9dd9-dbcdc7349f93.png)

ethman commented:
Okay, I just pushed some changes to your version, Andreas. Mostly touching up the docs, but you also missed propagating a handful of parameters which I fixed. 

At this point, my only remaining issue is with the implementation `_apply_user_ordered_effects()`. This is quite complicated and I cannot figure out how this is supposed to work in case it breaks in the future. I think a simpler and more interpretable solution might be to keep two indices around that define the start and end of the effects in the chain that are currently queued up for processing. Something like this:

lead_idx = 0
follow_idx = 0
while lead_idx <= len(fx_chain):
   if fx_chain[lead_idx] == fx_chain[follow_idx] or lead_idx == len(fx_chain):
      # same fx source, keep advancing the leader
      lead_idx += 1 
   else:
     # new fx source, apply effects and set follower to leader

abugler commented:
Yeah, I admit that function is very hacky. I'll try to clean it as much as possible!

ethman commented:
Hey Andreas, can I ask you to add a helper function to convert the `AudioSignal.effects_applied` list of strings to something we can use to reproduce the signal chain? So something along the lines of `[{'effect_name': 'fx1', 'effect_kwargs': {'param1': 0.1, ...}}, ...]`, that  parses the output of `FilterFunction.__str__()`. What I want is a list of things that we can pass to `AudioSignal.make_effect()` in a loop to completely apply the same effects more than once. You can put it in effects.py. Can you also write a test for the helper function?

Other than that, can you make `test_order()` more robust? Like make sure that the effects chain is applied in the way that we expect rather than just checking if the audio is not the same.

Sorry for this ask, I was going to do it myself but I won't have the bandwidth.

abugler commented:
No worries, I'll figure that out ASAP!

abugler commented:
> Hey Andreas, can I ask you to add a helper function to convert the `AudioSignal.effects_applied` list of strings to something we can use to reproduce the signal chain? So something along the lines of `[{'effect_name': 'fx1', 'effect_kwargs': {'param1': 0.1, ...}}, ...]`, that parses the output of `FilterFunction.__str__()`. What I want is a list of things that we can pass to `AudioSignal.make_effect()` in a loop to completely apply the same effects more than once. You can put it in effects.py. Can you also write a test for the helper function?
> 
> Other than that, can you make `test_order()` more robust? Like make sure that the effects chain is applied in the way that we expect rather than just checking if the audio is not the same.
> 
> Sorry for this ask, I was going to do it myself but I won't have the bandwidth.

I think it would actually be easier instead to store the `FilterFunction` objects as a list in `AudioSignal.effects_applied`.  Then instead, make `self._effects_chain` into a public variable, so you could do something like this:

```
signal2.effects_chain = signal1.effects_applied
```

Or for iteration:
```
for effect in signal1.effects_applied:
    filter_ = effect.filter
    params = effect.params
    signal2.make_effect(filter_, **params)
```

ethman commented:
Yeah, I think that's a great idea! Keep the `FilterFunction.__str__()` around because I'm sure that'll be useful. And also be sure to update AudioSignal constructor docs and add a sentence in `apply_effects()` about what's in the `effects_applied` list and how to use it (include your code snippet from your comment above).

Great work!

abugler commented:
So here are what I believe to be my final changes:
 + `effects_applied` and `effects_chain` are now properties of AudioSignal, and when you do `signal.effects_applied` or `signal.effects_chain`, it will result a **copy** of the desired list. This is to prevent madness like having two audio signals having the same list as an effects chain, where doing `signal.tremolo(.5, 6)` would add a tremolo effect to both signals. Sadly, it means this is not a valid operation:
```
signal2.effects_chain = signal1.effects_applied
```
 + `effects_applied` and `effects_chain` are both lists of `FilterFunction` objects. Since I did not originally plan on FilterFunction's being a class that user interact with, I cleaned up how the parameters are handled and stored, so that `FilterFunction.params` returns something helpful. Many input variables in FFmpeg are single letters, and that is originally how parameters where stored, so this has been changed too. 
+ You can now do this:
```
for effect in signal1.effects_applied:
    filter_ = effect.filter
    params = effect.params
    signal2.make_effect(filter_, **params)
```

These changes will be in the next commit :)

pseeth commented on docs/changelog.rst:
```diff
@@ -4,6 +4,7 @@ Changelog
 v1.1.1
 ------
 - Allowed MixSourceFolder to mix sources on the fly.
+- Added effects utilities.
```
Make this more descriptive. Also add another line about fixing the bug in Ignite.

pseeth commented on nussl/core/audio_signal.py:
```diff
@@ -688,7 +697,8 @@ def log_magnitude_spectrogram_data(self):
         (:obj:`np.ndarray`): Returns a real valued ``np.array`` with log magnitude spectrogram data.
         
-        The log magnitude spectrogram is defined as 20*log10(Abs(STFT)). Same shape as :attr:`stft_data`.
+        The log magnitude spectrogram is defined as 20*log10(Abs(STFT)).
```
Just noticing this now, looks a bit odd. Change to `20 * log10(abs(stft))`.

pseeth commented on nussl/core/audio_signal.py:
```diff
@@ -705,6 +715,28 @@ def log_magnitude_spectrogram_data(self):
         return 20 * np.log10(np.abs(self.stft_data) + 1e-8)
+    
+    @property
+    def effects_chain(self):
```
Please add a note indicating how you CANNOT use these properties, per our offline discussion.

pseeth commented on nussl/core/audio_signal.py:
```diff
@@ -705,6 +715,28 @@ def log_magnitude_spectrogram_data(self):
+        return self._effects_chain.copy()
+
+    @property
+    def effects_applied(self):
```
Add a note here as well, see comment on `effects_chain`.

pseeth commented on nussl/core/constants.py:
```diff
@@ -95,3 +96,10 @@
+##### augmentation.py
+# These values are found in the ffmpeg documentation in filters
+# that use the level_in argument, however, not all filters that
+# use this state these bounds in documentation with pytest.raises(ValueError):
```
Why is there a stray `pytest.raises` here? I don't fully understand this comment.

pseeth commented on nussl/datasets/transforms.py:
```diff
@@ -3,6 +3,7 @@
 import logging
 import random
 from collections import OrderedDict
+import librosa
```
Why is there a new import here? I don't see where librosa is called later on. I think this is a stray import that you don't need anymore.

pseeth commented on nussl/datasets/transforms.py:
```diff
@@ -11,6 +12,7 @@
 from sklearn.preprocessing import OneHotEncoder
 
 from .. import utils
+from ..core import AudioSignal
```
Don't think you need this import either.

pseeth commented on requirements.txt:
```diff
@@ -14,5 +14,9 @@ norbert
+pyloudnorm
+sox==1.4.0b0
+numba==0.48
+soxbindings
```
The `soxbindings` should move to a special way to install `nussl`: `pip install nussl[soxbindings]`, or `pip install -e ".[soxbindings]"` if installing from source. Do this in `setup.py`.

ethman commented:
Merging!

pseeth commented:
This is now in nussl 1.1.1. Thanks @abugler! Great work!
